t/21-needles.t

```diff
@@ -168,4 +179,75 @@ subtest 'handling relative paths in update_needle' => sub {
     };
 };
 
+subtest 'controller->_determine_needles_dir_for_job' => sub {
+    my $controller = OpenQA::WebAPI::Controller::Step->new;
+    $controller->app($t->app);
+    my %settings2 = %settings;
+    $settings2{CASEDIR} = 'https://something#fragment';
+    my $job = $schema->resultset('Jobs')->create_from_settings(\%settings2);
+    $job->create_result_dir;
+
+    subtest 'checkout_needles_sha = no' => sub {
+        $controller->app->config->{'scm git'} = {checkout_needles_sha => 'no'};
+        my ($needle_dirs, $needles_var) = $controller->_determine_needles_dir_for_job($job);
+        is($needle_dirs, undef, 'needles directory determined');
+        is($needles_var, undef, 'needles var determined');
+    };
+
+    subtest 'checkout_needles_sha = yes' => sub {
+        $controller->app->config->{'scm git'} = {checkout_needles_sha => 'yes'};
+        my ($needle_dirs, $needles_var) = $controller->_determine_needles_dir_for_job($job);
+        like($needle_dirs, qr|t/data/openqa/share/tests/fedora/needles$|, 'needles dirs is defined');
+        is($needles_var->value, 'https://something#fragment', 'needles var is defined');
+    };
+};
+
+subtest 'controller->_create_tmpdir_for_needles_refspec' => sub {
+    my $controller = OpenQA::WebAPI::Controller::Step->new;
+    $controller->app($t->app);
+    setup_fullstack_temp_dir('21-needles');
+    $controller->app->config->{'scm git'} = {checkout_needles_sha => 'yes'};
+
+    subtest 'without fragment and needles_git_hash in vars' => sub {
+        my %settings2 = %settings;
+        $settings2{CASEDIR} = 'https://something';
+        my $job = $schema->resultset('Jobs')->create_from_settings(\%settings2);
```

Comment:
perl has this feature where you can modify hash entries only for a certain scope:
```suggestion
        local $settings{CASEDIR} = 'https://something';
        my $job = $schema->resultset('Jobs')->create_from_settings(\%settings);
```
It saves just one line, but maybe you like it